File: linked_list/445_add_two_numbers_second_2-5-CtCI/Solution.java
import java.util.Stack;

/**
 You are given two non-empty linked lists representing two non-negative integers. The most significant digit comes first and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
 You may assume the two numbers do not contain any leading zero, except the number 0 itself.
 Follow up:
 What if you cannot modify the input lists? In other words, reversing the lists is not allowed.

 Example:
 Input: (7 -> 2 -> 4 -> 3) + (5 -> 6 -> 4)
 Output: 7 -> 8 -> 0 -> 7
 */
public class Solution {

    /**
     * Mine recursive solution. The idea is to add zeros to shorter number, like 84 + 09,
     * and then recursively add numbers beginning from the end
     * Space: O(max(M, N) + 1 + (M +N) )=O(max(M, N) + M + N)   - (M+N) - for the each level of recursion, max(M, N)+1 - length of the new list
     * Time: O(2max(M, N))=O(max(M, N))
     * */
    public ListNode addTwoNumbers(ListNode t1, ListNode t2) {
        if(t1 == null || t2 == null) return null;

        ListNode dummy = new ListNode(0), curr1 = t1, curr2 = t2;
        ListNode dHead = dummy;
        boolean firstShorter = false;

        while (curr1 != null || curr2 != null){
            if(curr1 == null || curr2 == null){
                dummy.next = new ListNode(0);
                dummy = dummy.next;
                if(curr1 == null) firstShorter = true;
            }

            if(curr1 != null) curr1 = curr1.next;
            if(curr2 != null) curr2 = curr2.next;
        }

        if(firstShorter && dHead.next != null) {
            dummy.next = t1;
            t1 = dHead.next;
        }
        else if(dHead.next != null){
            dummy.next = t2;
            t2 = dHead.next;
        }

        return addRecursive(t1, t2);
    }

    private ListNode addRecursive(ListNode t1, ListNode t2){
        ListNode sumNode = addRecursiveInternal(t1, t2);
        ListNode head = sumNode;
        if(sumNode.val >= 10){
            head = new ListNode(sumNode.val / 10);
            sumNode.val %= 10;
            head.next = sumNode;
        }
        return head;
    }

    private ListNode addRecursiveInternal(ListNode t1, ListNode t2){
        if(t1.next == null && t2.next == null){
            return new ListNode(t1.val + t2.val);
        }

        ListNode nextNode = addRecursiveInternal(t1.next, t2.next);
        int rest = nextNode.val / 10;
        nextNode.val %= 10;
        ListNode sumNode = new ListNode(rest + t1.val + t2.val);
        sumNode.next = nextNode;
        return sumNode;
    }



    /**
     * Not mine solution. Cool solution using stack
     * Space: O(M+N+max(M, N)) - two stacks + result list
     * Time: O(max(M, N) + M + N)
     * */
    public ListNode addTwoNumbers2222(ListNode l1, ListNode l2) {
        Stack<Integer> s1 = new Stack<>(), s2 = new Stack<>();

        for(; l1 != null; l1 = l1.next)
            s1.push(l1.val);

        for(; l2 != null; l2 = l2.next)
            s2.push(l2.val);

        ListNode res = new ListNode(0);
        int carry = 0;
        while(!s1.empty() || !s2.empty() || carry != 0) {
            ListNode cur = new ListNode(carry);
            if(!s1.empty()) cur.val += s1.pop();
            if(!s2.empty()) cur.val += s2.pop();
            carry = cur.val / 10;
            cur.val %= 10;
            cur.next = res.next;
            res.next = cur;
        }

        return res.next;
    }
}

File: linked_list/445_add_two_numbers_second_2-5-CtCI/Solution.py
# Sum Lists: You have two numbers represented by a linked list, where each node contains a single
# digit. The digits are stored in reverse order, such that the 1 's digit is at the head of the list. Write a
# function that adds the two numbers and returns the sum as a linked list.

# EXAMPLE
# Input: (7-> 1 -> 6) + (5 -> 9 -> 2) .Thatis,617 + 295.
# Output: 2 - > 1 - > 9. That is, 912.
# FOLLOW UP
# Suppose the digits are stored in forward order. Repeat the above problem.
# Input: (6 -> 1 -> 7) + (2 -> 9 -> 5).Thatis,617 + 295.
# Output: 9 - > 1 - > 2. That is, 912.

from typing import List



# 9 
# 8 -> 8 -> 2 

# 1 -> 2
# 8 -> 8 -> 2 

class Node:
    def __init__(self, val) -> None:
        self.next = None
        self.val = val

n1 = Node(9)
n2 = Node(9)

n3 = Node(8)
n4 = Node(8)
n3.next = n4

n5 = Node(1)
n5.next = Node(2)
################################################################################################

# My solution for initial problem, O(N) time
def sumLists(numb1: Node, numb2: Node) -> Node:
    rest = 0
    dumb = Node(0)
    head = dumb
    while (numb1 != None or numb2 != None):
        val1 = numb1.val if numb1 != None else 0
        val2 = numb2.val if numb2 != None else 0
        val = val1 + val2 + rest
        rest = int(val / 10)
        head.next = Node(val % 10)
        head = head.next
        if (numb1 != None): numb1 = numb1.next
        if (numb2 != None): numb2 = numb2.next
    
    if (rest > 0):
        head.next = Node(rest)
    return dumb.next

def sumListsForward(numb1: Node, numb2: Node) -> Node:
    first_len = getLength(numb1)
    second_len = getLength(numb2)

    tail, rest = 0, None
    if (first_len >= second_len):
        tail, rest = dfs(numb1, first_len, numb2, second_len)
    else:
        tail, rest = dfs(numb2, second_len, numb1, first_len)

    if (rest > 0):
        tmp = Node(rest)
        tmp.next = tail
        tail = tmp
    return tail

def getLength(node: Node) -> int:
    count = 0
    while (node != None):
        count += 1
        node = node.next
    return count 

def dfs(numb1: Node, len1: int, numb2: Node, len2: int) -> List[any]: # [Node, int rest]
    tail, rest, val = None, 0, 0
    if (len1 > len2):
        tail, rest = dfs(numb1.next, len1 - 1, numb2, len2)
        val = numb1.val + rest
    else:
        if (numb1.next == None and numb2.next == None):
            val = numb1.val + numb2.val
            return [Node(val % 10), int(val / 10)]
        
        tail, rest = dfs(numb1.next, len1 - 1, numb2.next, len2 - 1)
        val = numb1.val + numb2.val + rest

    tmp = Node(val % 10)
    tmp.next = tail
    return [tmp, int(val / 10)]

    

################################################################################################

def printList(head: Node) -> None:
    node = head
    values = []
    while (node != None):
        values.append(node.val)
        node = node.next
    print("->".join(str(val) for val in values))

#printList(sumLists(n1, n2)) # 9 + 9 = 18
#printList(sumLists(n1, n3)) # 9 + 88 = 97
#printList(sumLists(n3, Node(0))) # 88 + 0 = 88
#printList(sumLists(Node(1), Node(1))) # 1 + 1 = 2
#printList(sumLists(Node(5), Node(5))) # 5 +5 = 10


#printList(sumListsForward(n1, n2)) # 9 + 9 = 18
#printList(sumListsForward(n1, n3)) # 9 + 88 = 97
printList(sumListsForward(n5, n3)) # 12 + 88 = 100
printList(sumListsForward(Node(0), Node(0))) # 0 + 0 = 0
printList(sumListsForward(Node(0), Node(1))) # 0 + 1 = 1